### DISCUSSION TITLE
Internationalizing Copy and Paste

### DISCUSSION THREAD

**Discussion by chschnell** [copy/v86#1159] [Show and tell]
While working on graphical text mode I stumbled upon Copy and Paste which are still both tied to the en-US locale. It took a bit, but I think I here's how to internationalize these two functions.

Internationalizing Copy is relatively simply, but somewhere in the V86 settings the user will have to manually configure the 8-Bit codepage that is in use by the guest OS. I have published [codepage tables](https://github.com/chschnell/v86-i18n/tree/main/codepage-tables) here before.

Internationalizing Paste is a different beast though, and it is what this post is about.

I think the only proper way to implement a Paste function is by emulating key strokes on the keyboard, as is already the case in KeyboardAdapter.simulate_char() but without support for non-US keyboards. What I believe is missing is a "keyboard charset", that is the set of Unicode codepoints that can be generated by a specific keyboard (without resorting to Alt+3*Digit) and the mapping of scancodes (keystrokes) that produce them.

The website https://kbdlayout.info/ documents 217 different keyboard layouts in various formats, amongst them are KLC text files. A KLC text file is the output of Microsoft's Keyboard Layout Creator (MSKLC) encoded in Unicode, for example see the KLC code of the [US-keyboard](https://kbdlayout.info/kbdusx/download/klc). Unfortunately, the KLC file format is not documented, but it is practically self-explaining. It's possible to reliably generate the "keyboard charset" of any keyboard given its KLC file.

So I wrote a pyhton script [import_kbd.py](https://github.com/chschnell/v86-i18n/blob/main/keyboard-tables/import_kbd.py) that downloads a given set of KLC files, parses and extracts the keyboard charset, and then transforms that into a Javascript representation, see [keyboard_tables.js](https://github.com/chschnell/v86-i18n/blob/main/keyboard-tables/keyboard_tables.js) for the resulting output (contains mappings for 16 of the 217 known keyboards).

As that turned out to be slightly dense, I wrote a little HTML page that **[visualizes these mappings](https://chschnell.github.io/v86-i18n/keyboard-tables/)** for debugging. On the top you can select one out of 16 keyboards (could be extended to all 217 keyboards or any subset of it except for [CJK](https://en.wikipedia.org/wiki/CJK_characters)), on the left you see the repertoire of all typeable Unicode symbols and their scancode-mappings, and on the right you can enter some Unicode text into the top textarea that is transformed into the full scancode-sequence (including key-press and -release scancodes as well as handling dead keys) shown in the textarea below.

See method `Keyboard.text_to_scancodes(text)` in **[keyboard_lib.js](https://github.com/chschnell/v86-i18n/blob/main/keyboard-tables/keyboard_lib.js)** for the transformation from a Unicode text to its PC scancode sequence including modifer keys (Shift, Ctrl, Alt, AltGr), it's straight-forward.

Integrating this into V86 simply means to pump the produced scancode sequence into the V86 bus using `keyboard-code` events.

I included this in my [V86 demo](https://www.rekoba.de/~cs/v86/demo/), just select `Window -> Paste` from the menu. Here's a screenshot of the 149 (minus a few) German keyboard codepoints that I pasted into EDIT.EXE under guest OS FreeDOS:

![full-charset-de-DE](https://github.com/user-attachments/assets/349d8e87-3a09-45ff-84b9-49a992f79911)

I've pasted whole source code files into FreeDOS, it's stable and I believe that already says a lot, yet I still have to do a lot of testing under other OSes.

Like manually configuring the active codepage for Copy, a user would have to manually configure the active keyboard for Paste.

It's noteworthy that emulating keystrokes in this way does not require any knowledge of the codepage used by the guest OS, it's a straight Unicode-to-Scancode mapping.

As keyboard layouts are defined by international standards, all OSes should treat them alike (so it shouldn't matter that I source my infos from Windows). Yet, nothing is fixed, OSes typically support user-defined keyboard layouts, so all that can be supported here is constrained to international standards.

EDIT: Updated links.

EDIT 2: Updated links again.

**Comment by chschnell** [comment]
It works under Debian, too:

[debian-paste-de-charset.webm](https://github.com/user-attachments/assets/048e74b4-8f7c-413d-9f87-87d8cbdfb8c5)

**Comment by copy** [comment]
This looks interesting, thanks. One suggestions I would have is to compress layout information, by only specifying the keys that are different from the en-us layout.

**Reply by chschnell** [reply]
I packed the Codepoints of non-US keyboards that have identical keystroke-combinations like the US-keyboard. I also removed any mappings for codepoints < 32 (ASCII Space) except for 9 (Tab) and 10 (Enter), control characters don't make much sense to me in this context except for Tab and Enter.

In all that shaved off about ~~29%~~ 38% of the keyboard definition file, yay!

For comparison (unsurprisingly, packing is really effective for the UK-keyboard "kbduk"):

- [unpacked keyboard_charsets.js](https://www.rekoba.de/~cs/v86/keyboard-charset/keyboard_charsets_full.js) [50330 bytes]
- [packed keyboard_charsets.js](https://www.rekoba.de/~cs/v86/demo/include/keyboard_charsets.js) ~~[35823 bytes]~~ [31098 bytes]

Independent of that, I came across an issue I need to bring up.

At first I thought I'm doing something wrong, but when taking a closer look last night it became clear that I must throttle when sending scancodes across the bus to `PS2.kbd_send_code()`, as most of my synthetic keystrokes get lost if I don't.

`PS2.kbd_send_code()` puts the given byte into a ring buffer with a capacity of 1024 bytes which drops new data on overflow. Ok, but even if I send way less than 1024 bytes I experience loss, like already at 32 bytes or so. To be absolutely sure I resized that ring buffer to 1 Mb, same result.

I'm not sure what's going on here. I see it's correctly implemented with IRQ 01h + Port 60h/64h. Are we triggering IRQs faster than the guest OS can digest? Has anyone noticed this defect before?

I currently send at most 20 bytes per 100ms to the bus, but that's just a "works-for-me" solution. Unless this is a bug, I think the most robust way to handle this would be to go idle (setTimeout(0)) after each byte sent to the bus, even though that looks like 300 baud.

Sometimes my mouse goes whacky under V86 (jumping all over the place and I don't know how to provoke it), could that be the same issue? I wonder because the PS2 mouse implementation is so similar to the PS2 keyboard.

**Reply by copy** [reply]
> I'm not sure what's going on here. I see it's correctly implemented with IRQ 01h + Port 60h/64h. Are we triggering IRQs faster than the guest OS can digest? Has anyone noticed this defect before?

Yes, I'm aware of this issue. It's likely a bug in v86's ps2 or pic implementation, but I've never looked into it very deeply.

**Reply by chschnell** [reply]
> It's likely a bug in v86's ps2 or pic implementation, but I've never looked into it very deeply.

I think the problem is rooted outside of V86. A simple test with PS2.js only:

- inject 64 scancode bytes into PS2's keyboard buffer using [PS2.kbd_send_code()](https://github.com/copy/v86/blob/58d43361499d65e24295b20cb3857f123e261337/src/ps2.js#L270C1-L270C28), and
- monitor scancodes being eventually delivered to the guest OS in read-access handler of port 60h [PS2.port60_read()](https://github.com/copy/v86/blob/58d43361499d65e24295b20cb3857f123e261337/src/ps2.js#L401)

Input used for this test (each line 7 character plus 1 newline):

    1234567
    890abcd
    efghijk
    lmnopqr

Since each character is represented by 2 scancodes (press and release) this gives 64 bytes of scancode data:

    02 82 03 83 04 84 05 85 06 86 07 87 08 88 1C 9C
    09 89 0A 8A 0B 8B 1E 9E 30 B0 2E AE 20 A0 1C 9C
    12 92 21 A1 22 A2 23 A3 17 97 24 A4 25 A5 1C 9C
    26 A6 32 B2 31 B1 18 98 19 99 10 90 13 93 1C 9C

Now, while PS2.port60_read() is called exactly 64 times and returns the exact sequence of scancodes that was initially injected (!), the guest OS (Bochs+FreeDOS) only shows:

    1234567
    890abcd_

The cursor (underscore) stops at the end of the 2nd line, so the loss sets in after the 31st byte (or 30th) was delivered to the guest, yet the guest is requesting further bytes by reading port 60h and then they get lost. If the loss had set in after the 32nd byte than the cursor would be located on the next line. This happens every single time I tested.

I assume that when port 60h is being read it implies that the IRQ handshake must have been ok.

It looks as if whoever reads port 60h (BIOS, OS or application) is discarding the keyboard input, or am I missing something?

**Reply by chschnell** [reply]
On a second thought, whoever (BIOS, OS or application) reads will do that in an interrupt handler with an internal buffer of fixed size, and I guess that is where it overflows.

Anyway I've settled for a maximum of 15 scancode bytes per 100ms, that seems to work robustly.

**Reply by SuperMaxusa** [reply]
In Linux source code said "Expected maximum internal i8042 buffer size equals to 16 bytes" ([`drivers/input/serio/i8042.h#L42-L47`](https://github.com/torvalds/linux/blob/6db6a19f1ae31e29857b0f6f3e3896c22543be21/drivers/input/serio/i8042.h#L42-L47)), also confirmed by SeaBIOS sources: ([`src/hw/ps2port.c#L25`](https://github.com/coreboot/seabios/blob/2424e4c04aa30d90e85073ea41d18a7845460783/src/hw/ps2port.c#L25)). 

However, Linux doesn't seem to get this bug, personally I have only caught it in *-DOS systems, maybe something wrong with the buffer flush?

**Reply by chschnell** [reply]
Thanks for this!

I followed a bit into the Linux keyboard handler (I assume that it overrides the BIOS' handler). I'm using Debian 12 (32 bit) for my tests.

The main file is [i8042.c](https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c), it [registers](https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c#L1504) a [handler](https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c#L526) for IRQ 1 which is the keyboard IRQ. The IRQ handler reads a single byte from port 60h (keyboard data port) and forwards it to the serio subsystem [here](https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c#L610).

So up to this point this code does exactly what is expected, but I don't know a thing about that serio subsystem, I must assume it provides a much larger buffer.

While DOS is happy with a maximum of 15 scancode bytes (a "burst" of 15 bytes) per 100ms, Linux accepts way more, but it's not unlimited.

Pasting the source code of [keyboard_lib.js](https://github.com/chschnell/v86-i18n/blob/main/keyboard-tables/keyboard_lib.js) (16142 bytes of scancode data) with a maximum of 1024 bytes per burst works:

[debian-paste-1024.webm](https://github.com/user-attachments/assets/c18323e6-b8d6-47de-8abc-24d084de6db2)

But it overflows at 2048 bytes per burst:

[debian-paste-2048.webm](https://github.com/user-attachments/assets/8342c988-a82b-43c5-a40b-2541afb930b4)

So maximum burst size depends on the OS, I had hoped it wouldn't.

By the way, the past couple of days my keyboard code in keyboard_lib.js had a bug that surfaced only on international keyboards, it's fixed.

EDIT: Updated broken links.

**Reply by SuperMaxusa** [reply]
> So maximum burst size depends on the OS, I had hoped it wouldn't.

I have some thoughts that modern OSes (Linux, OpenBSD etc.) hooks keyboard IRQ themselves, while legacy OSes ([MS-DOS](https://github.com/microsoft/MS-DOS/blob/2d04cacc5322951f187bb17e017c12920ac8ebe2/v4.0-ozzie/bin/DISK2/BIOS/IBMDSK.ASM#L922-L937), FreeDOS, [bootOS](https://github.com/nanochess/bootOS/blob/b740e059d46e7072af836189370f0dac48c8f55c/os.asm#L502-L536) etc) uses a BIOS functions for input and this may be the reason for the different keyboard buffer sizes.

Here I compiled SeaBIOS with COM1 debug logging for testing: [seabios-com1debug.bin.gz](https://github.com/user-attachments/files/17134839/seabios-com1debug.bin.gz) ([config](https://github.com/user-attachments/files/17134874/config-seabios.txt)).

On QEMU, you can use `-bios` and `-serial stdio` options:
```
qemu-system-i386 -bios seabios-com1debug.bin -serial stdio <...>
```

On v86, you can use [xterm.js-based serial terminal](https://github.com/copy/v86/issues/203#issuecomment-753822011) (first run `make build/xterm.js` and load the library into your HTML file) and load debug BIOS in `V86` constructor.

--------------------------

For example, I tried to send 48 chars in MS-DOS and bootOS, it only shows the first 16 chars, but in SeaBIOS' log it receives all 48 commands:
```
i8042_command cmd=ae

i8042_wait_write

(47x)
```

Same behaviour on OpenBSD's boot phase (on UKC `boot>` prompt), but after loading it starts using own i8042 driver, so the problem goes away.

**Reply by chschnell** [reply]
> > So maximum burst size depends on the OS, I had hoped it wouldn't.
> 
> I have some thoughts that modern OSes (Linux, OpenBSD etc.) hooks keyboard IRQ themselves, while legacy OSes ([MS-DOS](https://github.com/microsoft/MS-DOS/blob/2d04cacc5322951f187bb17e017c12920ac8ebe2/v4.0-ozzie/bin/DISK2/BIOS/IBMDSK.ASM#L922-L937), FreeDOS, [bootOS](https://github.com/nanochess/bootOS/blob/b740e059d46e7072af836189370f0dac48c8f55c/os.asm#L502-L536) etc) uses a BIOS functions for input and this may be the reason for the different keyboard buffer sizes.
> 

That's 100% the case.

> Here I compiled SeaBIOS with COM1 debug logging for testing: [seabios-com1debug.bin.gz](https://github.com/user-attachments/files/17134839/seabios-com1debug.bin.gz) ([config](https://github.com/user-attachments/files/17134874/config-seabios.txt)).
> 
> On QEMU, you can use `-bios` and `-serial stdio` options:
> 
> ```
> qemu-system-i386 -bios seabios-com1debug.bin.gz -serial stdio <...>
> ```
> 
> On v86, you can use [xterm.js-based serial terminal](https://github.com/copy/v86/issues/203#issuecomment-753822011) (first run `make build/xterm.js` and load the library into your HTML file) and load debug BIOS in `V86` constructor.
> 

I didn't know this, very useful!

> For example, I tried to send 48 chars in MS-DOS and bootOS, it only shows the first 16 chars, but in SeaBIOS' log it receives all 48 commands:
> 
> ```
> i8042_command cmd=ae
> 
> i8042_wait_write
> 
> (47x)
> ```
> 
> Same behaviour on OpenBSD's boot phase (on UKC `boot>` prompt), but after loading it starts using own i8042 driver, so the problem goes away.

Yup.

I just wish the small bufferes weren't *that* small, bandwith is superlow because of that. Maybe two paste functions should be considered if this should get integrated: "slow paste" and "fast paste", but that seems a bit silly.

I will see if I can lower the 100ms pause between bursts a bit, but this will not change much about this.

**Reply by SuperMaxusa** [reply]
I have found that paste works correctly on MS-DOS 1.25 + Bochs BIOS (on previous and current versions of this demo), in other versions it skips a part of the keystrokes.

Just "for fun" I tried typing the Akalabeth for GWBASIC source code (https://nanochess.org/akalabeth.html), which is 29415 bytes and it tooked around 8 minutes. Unfortunately, on this BASIC in MS-DOS 1.25 the game got stuck on a level select, but this "weird" idea mostly worked.

https://github.com/user-attachments/assets/cb359d0c-bf8c-4972-8dbd-7d2a4f6635b2

**Reply by chschnell** [reply]
I love it! The time lapse is hilarious.

Actually this brings up old memories, 35 years ago we used to type these kind of games from printouts in PC magazines into our parent's 80286 or 80386 (can't remember), pages upon pages of code. Once a month a friend and I would pool our money and buy one of these, and then we'd had to meet for days to get it done, one was reading and checking while the other was typing. At least sometimes they had per-line checksums. Makes me complain less about the "slow" ~75 keys per second that we'll have to live with.

**Comment by chschnell** [comment]
I have made some progress and have a few preliminary results.

In order to make this a bit more presentable I moved all code into a new repository, see: **https://github.com/chschnell/v86-i18n**, the interactive keyboard demo is available there as a github page, see: https://chschnell.github.io/v86-i18n/keyboard-tables/. I've added example code to the repository.

The code has matured, I think I cannot make the representation of codepages and keyboard layouts any compacter than it is now, see [codepage_tables.js](https://github.com/chschnell/v86-i18n/blob/main/codepage-tables/codepage_tables.js) and [keyboard_tables.js](https://github.com/chschnell/v86-i18n/blob/main/keyboard-tables/keyboard_tables.js).

To summarize, I think this is what we learned so far:

1. when scancode data is placed into the PS/2 ringbuffer, a tight loop between V86 and the guest OS is initiated until the ringbuffer is exhausted (V86 raises INT 01h directly after the guest OS has retrieved a byte from the ringbuffer through port 60h, and the guest OS must handle the interrupt ASAP and then will read port 60h again triggering the next interrupt, and so on)
2. older PC operating systems are very constrained in their keyboard buffer (15 or 16 bytes per 100ms, unclear)
3. newer systems are less constrained (limit unclear), but V86 imposes a limit of 1024 bytes in its PS/2 ringbuffer

So unless the limit of V86's ringbuffer gets lifted we will always run into an upper bound, it just depends on the amount of pasted text.

I've also noticed that when I paste while the Firefox debugger is open (which slows V86 down considerably), scancode data will get lost.

@SuperMaxUSA noticed that some DOS/BIOS combinations may even be more constrained than for example FreeDOS/SeaBIOS which I use for tests (could you add which ones you tested?).

**Reply by SuperMaxusa** [reply]
Thanks for your info! I actually used Bochs Legacy BIOS for MS-DOS 1.25 because it has faster VGA text output than SeaBIOS in my eyes. About keyboard input, there are identical. While testing in MS-DOS, I found one interesting thing: keyboard input corruption can even depends on which program captures the keystrokes.

According to your previous tests, text input loses characters in FreeDOS EDIT, however, if VIm is used instead (you can click `Floppy -> Clone -> FreeDOS 1.3 720K Copy` in the demo and run `a:vim`), the text paste works fine.

Here is are some tests for this behavior, the original text is the first sentences of this post (literally):

As stated upper, in FreeDOS EDIT some characters are not sent, but some words can be recognized.

[freedos-edit.webm](https://github.com/user-attachments/assets/5bf45fd4-656a-4d39-bc7a-a78afd9a1df4)

In VIm (FreeDOS) and MS-DOS Editor works properly.

<details>
<summary>demo</summary>

[freedos-vim.webm](https://github.com/user-attachments/assets/f29658eb-d0b2-4c06-8632-18e79743298f)

[msdos-edit.webm](https://github.com/user-attachments/assets/0594df85-81dc-4d6e-8113-ead16763cdee)

(I split the text into 80 chars lines for MS-DOS Editor because it doesn't have "Text Wrap" option)

</details>

About BASIC, version 1.12 from MS-DOS 1.25 works properly, but GW-BASIC 2.02 from MS-DOS 2.11 skips part of text.

<details>
<summary>demo</summary>

1.25:

[msdos125-basic.webm](https://github.com/user-attachments/assets/70a7c96b-1989-474f-b27e-d64ce1b9f809)

2.11:

[msdos211-basic.webm](https://github.com/user-attachments/assets/07abdb40-39d8-44c8-9531-c7f4249c449b)

</details>

Last test is Windows 1.01 that very strangely leave only 42 characters and I don't know why, whereas on Windows 3.1 it worked properly.

[windows101-notepad.webm](https://github.com/user-attachments/assets/35264beb-2028-44ec-9e40-980ae3f43acf)

**Reply by chschnell** [reply]
Thanks for the extensive tests! Great effort, lots of new infos.

I'll repeat your tests and see what happens when pausing after each scancode (what I did initially), which is the slowest method to inject scancodes but of course also the most reliable one.
